From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Date: Fri, 13 May 2016 11:03:53 +0800
Subject: Order IPv6 nameservers before IPv4 for dns plugins (dnsmasq, bind)

IPv6 nameservers are usually ordered last because there's some chance of broken
routers dropping the AAAA requests before they reach the nameservers, but when
using dnsmasq, this is mitigated by how dnsmasq handles sending requests to the
nameservers. Since ordering IPv4 nameservers before IPv6 defeats the purpose of
enabling IPv6 and trying to prefer it whenever possible, make sure the IPv6
nameservers are actually showing up first in dns plugin configuration.

This is a very simple implementation of such re-ordering, just add the IPv6
nameservers first when building the list of IP configurations (which will
include nameservers, search domains, etc.), the IPv4 data will be appended
right after.

Bug-Ubuntu: https://bugs.launchpad.net/bugs/936712
---
 src/dns-manager/nm-dns-manager.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/dns-manager/nm-dns-manager.c b/src/dns-manager/nm-dns-manager.c
index 77e9221..9b7092c 100644
--- a/src/dns-manager/nm-dns-manager.c
+++ b/src/dns-manager/nm-dns-manager.c
@@ -782,14 +782,14 @@ build_plugin_config_lists (NMDnsManager *self,
 	 * still use the domain information in each config to provide split DNS if
 	 * they want to.
 	 */
-	if (priv->ip4_vpn_config)
-		*out_vpn_configs = g_slist_append (*out_vpn_configs, priv->ip4_vpn_config);
 	if (priv->ip6_vpn_config)
 		*out_vpn_configs = g_slist_append (*out_vpn_configs, priv->ip6_vpn_config);
-	if (priv->ip4_device_config)
-		*out_dev_configs = g_slist_append (*out_dev_configs, priv->ip4_device_config);
+	if (priv->ip4_vpn_config)
+		*out_vpn_configs = g_slist_append (*out_vpn_configs, priv->ip4_vpn_config);
 	if (priv->ip6_device_config)
 		*out_dev_configs = g_slist_append (*out_dev_configs, priv->ip6_device_config);
+	if (priv->ip4_device_config)
+		*out_dev_configs = g_slist_append (*out_dev_configs, priv->ip4_device_config);
 
 	for (iter = priv->configs; iter; iter = g_slist_next (iter)) {
 		if (!NM_IN_SET (iter->data, priv->ip4_vpn_config,
